Part1

hobwell opened the pull request.

hobwell committed:
Steps 1 to 9

updated replit files
implemented Point RValue constructor
Updated wrapper template to use forwarding
Updated wrapper template to have a print method
Templated a point specific implementation of the Wrapper
Implemented recursive variadic helper method
Implemented single argument variadic helper method

hobwell committed:
Added a missing universal reference

matkatmusic commented on main.cpp:
```diff
@@ -23,36 +23,76 @@ Make the following program work, which makes use of Variadic templates and Recur
-    Point& multiply(float m)
+    Point (float _x, float _y) : x (_x), y (_y) {}
+
+    Point (Point&& other) 
+    { 
+        x = other.x;
+        y = other.y;
+    }
```
consider if you need this constructor
- [x] revise

matkatmusic commented on main.cpp:
```diff
@@ -23,36 +23,76 @@ Make the following program work, which makes use of Variadic templates and Recur
 };
-
+
+template<>
+struct Wrapper<Point>
+{
+    Wrapper(Point&& p) : val (std::forward<Point>(p)) 
+    { 
+        std::cout << "Wrapper<Point>(" << typeid (val).name() << ")" << std::endl;
+    }
+
+    void print() const
+    {
+       std::cout << "Wrapper<Point>::print(" << val.toString() << ")" << std::endl;   
+    }
+
+    Point val;
+};
```
Project 9.1 Challenge 1:
There is a lot of repetition between the template class `Wrapper`, and the template specialization for `Wrapper<Point>`
Can you figure out how to eliminate the duplicate constructor?
hint: you only have to _explicitly specialize_ that which is unique.
You can specialize member functions without specializing the entire class :wink:
- [x] revise

matkatmusic commented on main.cpp:
```diff
@@ -66,10 +106,26 @@ struct Wrapper
  Wait for my code review.
  */
+
+template<typename T, typename ...Args>
+void variadicHelper (T&& first, Args&& ... remainingArgs)
+{
+    Wrapper (std::forward<T> (first)).print();
+    
+    if constexpr (sizeof ... (remainingArgs) > 0)
+        variadicHelper (std::forward<Args> (remainingArgs) ...);
+    
+}
+
+template<typename T>
+void variadicHelper (T&& only)
+{
+    Wrapper (std::forward<T> (only)).print();
+}
```
Project9.1 challenge 2:
There is another recursive solution.  it is possible for `Args ...` to be empty.
your task: remove your implementation for instruction `4)` and implement this alternate solution.
implement it **after** your existing `variadicHelper` function.
hint: what does the **recursive call** actually look like when called with an empty Variadic Parameter list
- [x] revise

hobwell committed:
Revisions

Removed unnecessary move constructor
Simplified Point specialization for Wrapper
updated recursive solution

matkatmusic commented on main.cpp:
```diff
@@ -23,36 +23,60 @@ Make the following program work, which makes use of Variadic templates and Recur
+
+    void print() const
+    {
+        if constexpr (std::is_same<double, Type>::value)
+        {
+            if (val == 3.5)
+                std::cout << "Wrapper::print(" << val << ")" << std::endl;
+        }
```
Why are you restricting the functionality of `print()` in this way?
- [x] revise

matkatmusic commented on main.cpp:
```diff
@@ -66,10 +90,26 @@ struct Wrapper
 int main()
-{
-    variadicHelper( 3, std::string("burgers"), 2.5, Point{3.f, 0.14f} );
+{
+    variadicHelper ();
+    variadicHelper (3, std::string ("burgers"), 2.5, Point{3.f, 0.14f});
```
Do not edit main()
- [x] revise

matkatmusic commented on main.cpp:
```diff
@@ -23,36 +23,60 @@ Make the following program work, which makes use of Variadic templates and Recur
+   std::cout << "Wrapper<Point>::print(" << val.toString() << ")" << std::endl;   
+}
+
+// Deduction Guide
+template <typename T>
+Wrapper(T&&) -> Wrapper<T>;
```
This is not needed
- [x] remove

matkatmusic commented on main.cpp:
```diff
@@ -23,36 +23,60 @@ Make the following program work, which makes use of Variadic templates and Recur
-    { 
-        std::cout << "Wrapper(" << typeid(val).name() << ")" << std::endl; 
-    }
+    Wrapper (Type&& t) : val (std::forward<Type> (t)) 
```
`move` was correct.
- [x] revise

hobwell committed:
Revisions

updated replit file to suppress -Wctad-maybe-unsupported warnings that appear in the absence of a deduction guide
add a comment to clarify why print was restricted
removed edit to main function

matkatmusic commented on main.cpp:
```diff
@@ -66,10 +93,25 @@ struct Wrapper
+{
+    Wrapper (std::forward<T> (first)).print();
+    
+    if constexpr (sizeof ... (remainingArgs) > 0)
+        variadicHelper (std::forward<Args> (remainingArgs) ...);
```
remove the `if constexpr` check
- [x] revise

matkatmusic commented on main.cpp:
```diff
@@ -23,36 +23,63 @@ Make the following program work, which makes use of Variadic templates and Recur
+    void print() const
+    {
+
+        // the first parameter of the variadic helper will be the same as Type, given how Wrapper is constructed, so we can check here if it is a double
+        if constexpr (std::is_same<double, Type>::value)
+        {
+            // then check if it is 3.5
+            if (val == 3.5)
+                std::cout << "Wrapper::print(" << val << ")" << std::endl;
+        }
```
remove the `if constexpr` check
- [x] revise

hobwell committed:
Revised, with commentary

hobwell committed:
Revused, cleared comments
